truss/templates/server/model_wrapper.py

```diff
@@ -202,144 +383,91 @@ def try_load(self):
+                # Even for async generators, don't await here.
+                return self._model.predict(*args)
+            if descriptor.is_async:
+                return await self._model.predict(*args)
+            # Offload sync functions to thread, to not block event loop.
+            return await to_thread.run_sync(self._model.predict, *args)
 
     async def postprocess(
         self,
-        response: Any,
-    ) -> Any:
-        # Similar to the predict function, it is possible for postprocess
-        # to return either a generator or async generator, in which case
-        # just return the generator.
-        #
-        # It can also return a coroutine or just be a function, in which
-        # case either await, or offload to a thread respectively.
-        if not hasattr(self._model, "postprocess"):
-            return response
-
-        if inspect.isasyncgenfunction(
-            self._model.postprocess
-        ) or inspect.isgeneratorfunction(self._model.postprocess):
-            return self._model.postprocess(response)
-
-        if inspect.iscoroutinefunction(self._model.postprocess):
-            return await errors.intercept_exceptions(
-                self._model.postprocess, self._logger
-            )(response)
-
-        return await to_thread.run_sync(
-            errors.intercept_exceptions(self._model.postprocess, self._logger), response
-        )
-
-    async def write_response_to_queue(
-        self, queue: asyncio.Queue, generator: AsyncGenerator, span: trace.Span
-    ):
+        result: Union[serialization.InputType, Any],
+        request: starlette.requests.Request,
+    ) -> serialization.OutputType:
+        # The postprocess function can handle outputs of `predict`, but not
+        # generators and responses - in that case predict must return directly
+        # and postprocess is skipped.
+        # The result type can be the same as for predict.
+        descriptor = self.model_descriptor.postprocess
+        if not descriptor:
+            return result
+
+        args = ArgConfig.prepare_args(descriptor, result, request)
+        with errors.intercept_exceptions(self._logger):
+            if descriptor.is_async:
+                return await self._model.postprocess(*args)
+            # Offload sync functions to thread, to not block event loop.
+            return await to_thread.run_sync(self._model.postprocess, *args)
+
+    async def _write_response_to_queue(
+        self,
+        queue: asyncio.Queue,
+        generator: AsyncGenerator[bytes, None],
+        span: trace.Span,
+    ) -> None:
         with tracing.section_as_event(span, "write_response_to_queue"):
             try:
                 async for chunk in generator:
-                    # TODO: consider checking `request.is_disconnected()` for
-                    #   client-side cancellations and freeing resources.
-                    await queue.put(ResponseChunk(chunk))
+                    await queue.put(chunk)
```

Comment:
why did we get rid of the ResponseChunk thing? If the generator yields `None` now, it will get dequeued and terminate the stream

Reply:
Good catch!

Although returning `None` would give an internal server error anyway, because response content can only be bytes/str.

This single source file used two different patterns for marking the end of iteration... this `ResponseChunk` (which appeared to me more like a leftover from some other ideas) and `FINAL_GENERATOR_VALUE`... I've unified this to `SENTINEL` now.